src/view.c

```diff
@@ -423,44 +423,64 @@ view_apply_unmaximized_geometry(struct view *view)
 	}
 }
 
+static void
+set_maximized(struct view *view, bool maximized)
+{
+	if (view->impl->maximize) {
+		view->impl->maximize(view, maximized);
+	}
+	if (view->toplevel_handle) {
+		wlr_foreign_toplevel_handle_v1_set_maximized(
+			view->toplevel_handle, maximized);
+	}
+	view->maximized = maximized;
+}
+
+/* Un-maximize, un-tile and move to specific geometry */
 void
-view_maximize(struct view *view, bool maximize)
+view_restore_to(struct view *view, struct wlr_box geometry)
+{
+	if (view->fullscreen) {
+		return;
+	}
+	if (view->maximized) {
+		set_maximized(view, false);
+	}
+	view->tiled = 0;
+	view_move_resize(view, geometry);
+}
+
+void
+view_maximize(struct view *view, bool maximize, bool store_natural_geometry)
 {
 	if (view->maximized == maximize) {
 		return;
 	}
 	if (view->fullscreen) {
 		return;
 	}
-	if (view->impl->maximize) {
-		view->impl->maximize(view, maximize);
-	}
-	if (view->toplevel_handle) {
-		wlr_foreign_toplevel_handle_v1_set_maximized(
-			view->toplevel_handle, maximize);
-	}
+	set_maximized(view, maximize);
 	if (maximize) {
-		interactive_end(view);
-		if (!view->tiled) {
+		interactive_end(view, /*allow_snap*/ false);
```

Comment:
I don't remember why we have to call `interactive_end()` here in the first place, if you do would you mind adding a comment here?

Reply:
Added comments.